File: ios/Classes/DetectionData.swift
//
//  DetectionWithTime.swift
//  flutter_webrtc
//
//  Created by MacBook 16 on 17.06.2022.
//


import Foundation
import Flutter


struct DetectionData {
    var frames: [String:[Square]]
    let aspect: Double
    let xSqCount: Int
    let ySqCount: Int
    let frameIntervalMs: Int
    var duration: Int = 0
    
    init(detectionFrame: DetectionFrame,
         frameIndex: Int,
         frameIntervalMs: Int) {
        self.frames = ["\(frameIndex)": detectionFrame.detectedList]
        self.aspect = detectionFrame.aspectRatio
        self.xSqCount = detectionFrame.xCount
        self.ySqCount = detectionFrame.yCount
        self.frameIntervalMs = frameIntervalMs        
    }
    
    mutating func addDetection(detection: DetectionFrame, frameIndex: Int) throws  {
        guard xSqCount == detection.xCount,
              ySqCount == detection.yCount,
              aspect == detection.aspectRatio else {
            throw NSError(domain: "DetectionData",
                          code: 1,
                          userInfo: ["message": "Detection data is not compatible"])
        }
        frames["\(frameIndex)"] = detection.detectedList
    }
    
    
    
    func toMap() -> [String: Any] {
        let framesJson = frames.mapValues({squares in squares.map({sq in sq.toString()})})
        return [
            "frames": framesJson,
            "a": NSNumber(value: aspect),
            "x": NSNumber(value: xSqCount),
            "y": NSNumber(value: ySqCount),
            "ft": NSNumber(value: frameIntervalMs),
            "d": NSNumber(value: duration)
        ]
    }
}

File: android/src/main/java/com/cloudwebrtc/webrtc/videoRecorder/DetectionData.kt
package com.cloudwebrtc.webrtc.videoRecorder

import android.util.Log
import com.cloudwebrtc.webrtc.detection.DetectionFrame

public data class DetectionData(
    val frames: MutableMap<String, List<String>>,
    val aspect: Double,
    val xCount: Int,
    val yCount: Int,
    val frameInterval: Int,
    var duration: Long = 0
) {

    fun toMap(): Map<String, Any> {
        return mapOf(
            "f" to frames,
            "a" to aspect,
            "x" to xCount,
            "y" to yCount,
            "i" to frameInterval,
            "d" to duration
        )
    }


    constructor(detectionResult: DetectionFrame, frameIndex: String, frameInterval: Int) : this(
        frames = mutableMapOf(frameIndex to detectionResult.detectedList.map { it.toString() }),
        aspect = detectionResult.aspectRatio,
        xCount = detectionResult.xCount,
        yCount = detectionResult.yCount,
        frameInterval = frameInterval
    )

    fun addFrame(frameIndex: String, detectionResult: DetectionFrame) {
        if (detectionResult.xCount != xCount || detectionResult.yCount != yCount) {
            Log.e("Motion detection","DetectionData: xCount or yCount not match")
            return
        }
        frames[frameIndex] = detectionResult.detectedList.map { it.toString() }
    }

}
